File: opspilot-enterprise/packages/shared-types/src/intent.ts
export type IntentDomain = "vmware" | "k8s" | "host" | "jenkins" | "knowledge" | "unknown";
export type RecoveryDecision = "recovered" | "clarify_required" | "rejected";
export type RiskLevel = "L0" | "L1" | "L2" | "L3" | "L4";
export type ExecutionIntentMode = "read" | "plan" | "execute";

export interface SlotValue {
  name: string;
  value: unknown;
  source: "user" | "memory" | "cmdb" | "tool_discovery" | "inferred";
  confidence: number;
}

export interface EvidenceRef {
  type: "session" | "knowledge" | "cmdb" | "tool_discovery";
  ref_id: string;
  summary: string;
  score: number;
}

export interface ScoreBreakdown {
  rules: number;
  entity_match: number;
  slot_completeness: number;
  memory_boost: number;
  llm_rerank: number;
  domain_gate_score: number;
  target_resolution_score: number;
}

export interface ResolutionRef {
  ref_id: string;
  name: string;
  type: string;
  matched_by: string;
  connection_id?: string | null;
  environment?: string | null;
  aliases: string[];
  score: number;
}

export interface IntentCandidate {
  intent_code: string;
  domain: IntentDomain;
  action: string;
  description: string;
  resource_scope: "single" | "multiple" | "cluster" | "global";
  environment?: string | null;
  memory_refs: string[];
  score: number;
  score_breakdown: ScoreBreakdown;
  slots: SlotValue[];
  missing_slots: string[];
  inferred_environment?: string | null;
  inferred_risk_level?: RiskLevel | null;
  evidence: EvidenceRef[];
  target_object_raw?: string | null;
  target_object_resolved?: string | null;
  target_type?: string | null;
  resolution_confidence: number;
  resolution_refs: ResolutionRef[];
}

export interface IntentRecoveryRun {
  run_id: string;
  conversation_id: string;
  user_id: string;
  channel: string;
  tenant_id?: string | null;
  raw_utterance: string;
  normalized_utterance: string;
  candidates: IntentCandidate[];
  chosen_intent?: IntentCandidate | null;
  decision: RecoveryDecision;
  clarify_reasons: string[];
  rejected_reasons: string[];
  created_at: string;
}

export interface ExecutionIntent {
  mode: ExecutionIntentMode;
  reason: string;
  target_tool?: string | null;
  guardrails: string[];
}

export interface RiskContext {
  environment: string;
  resource_scope: "single" | "multiple" | "cluster" | "global";
  object_count: number;
}

export interface IntentAnalyzeResponse {
  run_id: string;
  decision: RecoveryDecision;
  selected_intent?: IntentCandidate | null;
  candidates: IntentCandidate[];
  execution_intent: ExecutionIntent;
  risk_context: RiskContext;
  context_hints: Record<string, unknown>;
  normalized_utterance: string;
  memory_refs: string[];
  evidence_refs: string[];
  clarify_reasons: string[];
  rejected_reasons: string[];
  clarify_card?: Record<string, unknown> | null;
  approval_card?: Record<string, unknown> | null;
  rag_plan?: Record<string, unknown> | null;
  run: IntentRecoveryRun;
}

File: opspilot-enterprise/packages/shared-schema/src/opspilot_schema/intent.py
from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

IntentDomain = Literal["vmware", "k8s", "host", "jenkins", "knowledge", "unknown"]
RecoveryDecision = Literal["recovered", "clarify_required", "rejected"]
RiskLevel = Literal["L0", "L1", "L2", "L3", "L4"]
SlotSource = Literal["user", "memory", "cmdb", "tool_discovery", "inferred"]
EvidenceSourceType = Literal["session", "knowledge", "cmdb", "tool_discovery"]
ExecutionIntentMode = Literal["read", "plan", "execute"]


class SlotValue(BaseModel):
    name: str
    value: Any = None
    source: SlotSource = "user"
    confidence: float = 0.0


class EvidenceRef(BaseModel):
    type: EvidenceSourceType
    ref_id: str
    summary: str = ""
    score: float = 0.0


class ScoreBreakdown(BaseModel):
    rules: float = 0.0
    entity_match: float = 0.0
    slot_completeness: float = 0.0
    memory_boost: float = 0.0
    llm_rerank: float = 0.0
    domain_gate_score: float = 0.0
    target_resolution_score: float = 0.0


class ResolutionRef(BaseModel):
    ref_id: str
    name: str
    type: str
    matched_by: str = ""
    connection_id: str | None = None
    environment: str | None = None
    aliases: list[str] = Field(default_factory=list)
    score: float = 0.0


class IntentCandidate(BaseModel):
    intent_code: str
    domain: IntentDomain
    action: str
    description: str = ""
    resource_scope: Literal["single", "multiple", "cluster", "global"] = "single"
    environment: Optional[str] = None
    memory_refs: list[str] = Field(default_factory=list)
    score: float = 0.0
    score_breakdown: ScoreBreakdown = Field(default_factory=ScoreBreakdown)
    slots: list[SlotValue] = Field(default_factory=list)
    missing_slots: list[str] = Field(default_factory=list)
    inferred_environment: Optional[str] = None
    inferred_risk_level: Optional[RiskLevel] = None
    evidence: list[EvidenceRef] = Field(default_factory=list)
    target_object_raw: str | None = None
    target_object_resolved: str | None = None
    target_type: str | None = None
    resolution_confidence: float = 0.0
    resolution_refs: list[ResolutionRef] = Field(default_factory=list)


class IntentRecoveryRun(BaseModel):
    run_id: str
    conversation_id: str
    user_id: str
    channel: str = "web"
    tenant_id: Optional[str] = None
    raw_utterance: str
    normalized_utterance: str
    candidates: list[IntentCandidate] = Field(default_factory=list)
    chosen_intent: Optional[IntentCandidate] = None
    decision: RecoveryDecision = "clarify_required"
    clarify_reasons: list[str] = Field(default_factory=list)
    rejected_reasons: list[str] = Field(default_factory=list)
    created_at: str

    @property
    def top_candidates(self) -> list[IntentCandidate]:
        return self.candidates[:3]

    @property
    def missing_slots(self) -> list[str]:
        if self.chosen_intent:
            return self.chosen_intent.missing_slots
        if self.candidates:
            return self.candidates[0].missing_slots
        return []

    @property
    def evidence_summary(self) -> list[str]:
        evidence = self.chosen_intent.evidence if self.chosen_intent else (self.candidates[0].evidence if self.candidates else [])
        return [item.summary for item in evidence if item.summary]


class IntentRecoverInput(BaseModel):
    conversation_id: str
    user_id: str
    channel: str = "web"
    utterance: str
    tenant_id: Optional[str] = None
    history: list[dict[str, Any]] = Field(default_factory=list)
    memory: list[str] = Field(default_factory=list)
    resource_catalog: list[dict[str, Any]] = Field(default_factory=list)


class ExecutionIntent(BaseModel):
    mode: ExecutionIntentMode = "read"
    reason: str = ""
    target_tool: str | None = None
    guardrails: list[str] = Field(default_factory=list)


class RiskContext(BaseModel):
    environment: str = "unknown"
    resource_scope: Literal["single", "multiple", "cluster", "global"] = "single"
    object_count: int = 1


class IntentAnalyzeInput(IntentRecoverInput):
    ui_context: dict[str, Any] = Field(default_factory=dict)
    prefer_execute: bool | None = None


class IntentAnalyzeResponse(BaseModel):
    run_id: str
    decision: RecoveryDecision
    selected_intent: IntentCandidate | None = None
    candidates: list[IntentCandidate] = Field(default_factory=list)
    execution_intent: ExecutionIntent = Field(default_factory=ExecutionIntent)
    risk_context: RiskContext = Field(default_factory=RiskContext)
    context_hints: dict[str, Any] = Field(default_factory=dict)
    normalized_utterance: str
    memory_refs: list[str] = Field(default_factory=list)
    evidence_refs: list[str] = Field(default_factory=list)
    clarify_reasons: list[str] = Field(default_factory=list)
    rejected_reasons: list[str] = Field(default_factory=list)
    clarify_card: dict[str, Any] | None = None
    approval_card: dict[str, Any] | None = None
    rag_plan: dict[str, Any] | None = None
    run: IntentRecoveryRun
